File: Java/Populating_Next_Right_Pointers_in_Each_Node.java
/**
 * Problem Link: https://leetcode.com/problems/populating-next-right-pointers-in-each-node/
 * Binary Tree Level Traverse.
 * Definition for binary tree with next pointer.
 * public class TreeLinkNode {
 *     int val;
 *     TreeLinkNode left, right, next;
 *     TreeLinkNode(int x) { val = x; }
 * }
 */
public class Solution {
    public void connect(TreeLinkNode root) {
        if(root == null) return;
        Queue<TreeLinkNode> q = new LinkedList<TreeLinkNode>();
        q.offer(root);
        int cur = 1;
        int nxt = 0;
        while(!q.isEmpty()) {
            TreeLinkNode node = q.poll();
            cur--;
            if(node.left != null) {
                q.offer(node.left);
                nxt++;
            }
            if(node.right != null) {
                q.offer(node.right);
                nxt++;
            }
            if(cur == 0) {
                node.next = null;
                cur = nxt;
                nxt = 0;
            } else {
                node.next = q.peek();
            }
        }
    }
}

File: Swift/populating-next-right-pointers-in-each-node.swift
/**
 * https://leetcode.com/problems/populating-next-right-pointers-in-each-node/
 * 
 * 
 */ 
// Date: Tue May 26 12:03:32 PDT 2020
/**
 * Definition for a Node.
 * public class Node {
 *     public var val: Int
 *     public var left: Node?
 *     public var right: Node?
 *       public var next: Node?
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.left = nil
 *         self.right = nil
 *         self.next = nil
 *     }
 * }
 */

class Solution {
    func connect(_ root: Node?) -> Node? {
        guard let root = root else { return nil }
        var queue = [root]
        while queue.isEmpty == false {
            var nextQueue: [Node] = []
            for index in 0 ..< queue.count {
                if index == queue.count - 1 {
                    queue[index].next = nil
                } else {
                    queue[index].next = queue[index + 1]
                }
                if let left = queue[index].left {
                    nextQueue.append(left)
                }
                if let right = queue[index].right {
                    nextQueue.append(right)
                }
            }
            queue = nextQueue
            print("Count: \(queue.count)")
        }
        return root
    }
}
